Add spec for multisig actor wallet

whyrusleeping opened the pull request:
In order to provide a reasonable solution to holding filecoin at launch, we add a multisig account actor.

The design is modeled after the [gnosis multisig wallet](https://github.com/gnosis/MultiSigWallet/blob/master/contracts/MultiSigWallet.sol), but with only the bare minimum of features (no spend limits or anything funky).

Some design choices that may be worth discussing:

* Transactions are kept around after being executed or canceled, and cleared later by an explicit call to clear transactions. This is done to make auditing easier, seeing transaction history simply requires a lookup in the current state tree, as opposed to crawling back through the chain (being more friendly to lite clients). The alternative is to remove the transaction from the actor state as soon as it is finished. I'm on the fence about which option is best.

* Transactions may be canceled by any of the signers. This seems the simplest approach, though if a signer were to be 'rogue' then they could make it very difficult to spend any funds, or even remove them as a signer. This could be circumvented by asking a miner to mine their transactions all at the same time, preventing the rogue signer from cancelling it. One alternative is to just not have cancelations, and let the 'bad' transaction just hang out forever (bad solution IMO). Another would be to implement some 'required threshold' for canceling a request, this feels a little heavy to me but I could be convinced.

* Self modification methods (add/rm signer, change req) are implemented by doing a multisig transaction invoking the desired method on the contract itself. This means we only have to implement the 'signature threshold' logic in one place, and is generally simpler to reason about. If someone has strong feelings about this, i'm open to discuss.

dignifiedquire commented on actors.md:
```diff
@@ -478,6 +478,185 @@ func UpdatePeerID(pid PeerID) {
+    
+    tx.Approved.Append(msg.From)
+    
+    if len(tx.Approved) > self.Required {
```
shouldn’t this be >= ?

whyrusleeping replied:
good catch.

- [ ] Fix

dignifiedquire commented on actors.md:
```diff
@@ -478,6 +478,185 @@ func UpdatePeerID(pid PeerID) {
+        Fatal("add signer must be called by wallet itself")
+    }
+    if isSigner(signer) {
+        Fatal("new address is already an owner")
```
typo: already a signer

dignifiedquire commented on actors.md:
````diff
@@ -478,6 +478,185 @@ func UpdatePeerID(pid PeerID) {
+
+```go
+func ClearCompleted() {
+    if !isOwner(msg.From) {
````
what is the difference between owner and signer? is owner the creator of the wallte? can the ownership be changed?

whyrusleeping replied:
there is none, i was using the phrasing owner at one point, and switched it to signer. Signer is the appropriate thing to use here

dignifiedquire commented on actors.md:
````diff
@@ -478,6 +478,185 @@ func UpdatePeerID(pid PeerID) {
+```go
+func RemoveSigner(signer Address) {
+    if msg.From != self.Address {
+        Fatal("remove signer must be called by wallet itself")
````
this error message confuses me,

sternhenri replied:
+1

dignifiedquire commented on actors.md:
````diff
@@ -478,6 +478,185 @@ func UpdatePeerID(pid PeerID) {
+
+```go
+func ChangeRequirement(req int) {
+    if msg.From != self.Address {
````
how should this be called? is Address the general address of an actor?

whyrusleeping replied:
The way this works is you tell the wallet to send a multisig transaction to itself calling this method. that way we don't have to implement the signature threshold stuff differently for each of these changes.

sternhenri replied:
from reading the PR notes (not obvious in code itself) i am guessing the self.address is a well-signed threshold, ie m-of-n signers.

sternhenri commented on actors.md:
```diff
@@ -478,6 +478,185 @@ func UpdatePeerID(pid PeerID) {
+        Fatal("not authorized")
+    }
+    
+    tx := getTransction(txid)
```
typo

sternhenri commented on actors.md:
````diff
@@ -478,6 +478,185 @@ func UpdatePeerID(pid PeerID) {
+### AddSigner
+
+```go
+func AddSigner(signer Address) {
````
Forgot to check is function is called by a signer (or arguably should be by owner)

sternhenri commented on actors.md:
```diff
@@ -478,6 +478,185 @@ func UpdatePeerID(pid PeerID) {
+        Fatal("change requirement must be called by wallet itself")
+    }
+    
+    self.Required = req
```
need some sanity checks on the int, i.e. >= 1...

sternhenri commented:
This is very nifty. Numbering your design considerations, some thoughts to each:
1) I think this is fine indeed. How big an issue is this for lite clients though? It seems like you'd want to keep pretty little in memory and do the actual crawling just once...
2) You have adding and removing of signers done by the wallet address, i.e. by a threshold of signers... Why not apply the same rule to canceling transactions? It seems like the rogue signer issue devolves to a question of who gets to pick signers, so this approach would make sense to me... Make it an m-of-n thing as well.
3) +1

whyrusleeping commented:
Alright @dignifiedquire @sternhenri Now that the init actor PR has merged, i feel this is unblocked. I've rebased it, addressed the concerns raised, and cleaned up a few things. Give it another look over please!

nicola commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+## Multisig Account Actor
+
+A basic multisig account actor. Allows sending of messages like a normal account actor, but with the requirement of M of N parties agreeing to the operation. Any party can veto any operation (which is a potentially debateable design decision). Completed and/or cancelled operations stick around in the actors state until explicitly cleared out.
+
```
Meta note, this kind of multi-sig is ok, but since we use BLS signature already anyway, why don't we use threshold signatures instead? This would reduce multisig on-chain footprint. (I might be missing context on threshold sig vs on-chain multi-sig).

nicola commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+        Fatal("not authorized")
+    }
+
+    txid := self.NextTxID
```
where do we keep the nextTxID?

nicola commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+
+### Propose
+
+> Propose is used to propose a new transaction to be sent by this multisig. The proposer must be a signer, and the proposal also serves as implicit approval from the proposer. If only a single signature is required, then the transaction is executed immediately.
```
can you specify if this is happening on-chain or off-chain?

nicola replied:
If the answer is "on-chain", a lot of signature collection can definitely happen offchain.

nicola commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+
+### Approve
+
+> Approve is called by a signer to approve a given transaction. If their approval pushes the approvals for this transaction over the threshold, the transaction is executed.
```
on-chain or off-chain here too

nicola commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
 
 TODO
+
+## Multisig Account Actor
```
How do I create one of those actors?

whyrusleeping replied:
Using the init actor.

- [ ] detail how this works

nicola commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+
+## Multisig Account Actor
+
+A basic multisig account actor. Allows sending of messages like a normal account actor, but with the requirement of M of N parties agreeing to the operation. Any party can veto any operation (which is a potentially debateable design decision). Completed and/or cancelled operations stick around in the actors state until explicitly cleared out.
```
technically "any party can veto" is not correct. If you pick n of m and n approve and one veto, transaction gets exectured.

whyrusleeping replied:
lets go with "you can cancel your own transaction, or the transactions of anyone who is no longer a signer", and/or "any transaction that hasnt been approved in X blocks can be removed"

nicola commented on actors.md:
````diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+
+    tx.Canceled = true
+}
+```
````
Can you explain how this can happen?
This sounds like the "veto" power, if n of m required people sign, why would one of the remaining party have power to stop the transaction? this sounds like a setting where is not always desirable

sternhenri replied:
IF n have signed, transaction goes through, however someone can cancel ahead of then. This does mean that an evil signer can DoS all m-1 other parties, though the proposer can just remove them from the signer set and try again.

nicola commented on actors.md:
````diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+### SwapSigner
+
+```go
+func SwapSigner(old, new Address) {
````
shouldn't be this something approved by the required set of people?

nicola commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+
+
+
+### ChangeRequirement
```
shouldn't be this something approved by the required set of people?

dignifiedquire commented on actors.md:
````diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+    Completed bool
+    Canceled bool
+}
+```
````
It would be nice to have a method that queries the current state of a transaction, so I can know if I need/want to do sth about it

sternhenri replied:
On that note, rather than have 
`Completed bool
Canceled bool`

We might just have a State int, with a few potential states: ongoing, completed, canceled...

dignifiedquire commented on actors.md:
````diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+### AddSigner
+
+```go
+func AddSigner(signer Address) {
````
why does this not need approval from others?

dignifiedquire commented on actors.md:
````diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+
+```go
+func Cancel(txid uint64) {
+    if !isSigner(msg.From) {
````
why does this not need approval from others?

dignifiedquire commented on actors.md:
````diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+
+```go
+func Approve(txid uint64) {
+    if !isSigner(msg.From) {
````
`isSigner` is not defined

dignifiedquire commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+        Fatal("not authorized")
+    }
+
+    tx := getTransaction(txid)
```
`getTransaction` is not defined

whyrusleeping commented:
TODO: add roughly the following to the spec: "Self modification methods (add/rm signer, change req) are implemented by doing a multisig transaction invoking the desired method on the contract itself. This means we only have to implement the 'signature threshold' logic in one place, and is generally simpler to reason about."

sternhenri commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+func Propose(to Address, value TokenAmount, method string, params []byte) uint64 {
+    if !isSigner(msg.From) {
+        Fatal("not authorized")
+    }
```
Shouldn't you conceivably be allowed to propose a transaction for which you are not a signer?

whyrusleeping replied:
I don't think thats useful. Then any random person can spam your multisig with bad transactions that we have to clear out, and can be misleading (making a typo on an approval transaction much more risky)

sternhenri commented on actors.md:
```diff
@@ -708,3 +708,233 @@ func UpdatePeerID(pid PeerID) {
+func Cancel(txid uint64) {
+    if !isSigner(msg.From) {
+        Fatal("not authorized")
+    }
```
Alternative would be splitting out a `isProposer` permission so only proposer can cancel, as opposed to any signer. I don't think it matters though, somewhat equivalent.

sternhenri commented:
Understood and agreed

On Tue, Mar 26 2019 at 12:18, < notifications@github.com > wrote:

> 
> 
> 
> *@whyrusleeping* commented on this pull request.
> 
> 
> 
> In actors.md (
> https://github.com/filecoin-project/specs/pull/139#discussion_r269274406 )
> :
> 
> > +### Propose + +> Propose is used to propose a new transaction to be
> sent by this multisig. The proposer must be a signer, and the proposal
> also serves as implicit approval from the proposer. If only a single
> signature is required, then the transaction is executed immediately. + +|
> Name | Type | Description | +| --- | --- | --- | +| `to` | `Address` | The
> address of the target of the proposed transaction. | +| `value` |
> `TokenAmount` | The amount of funds to send with the proposed transaction
> | +| `method` | `string` | The method that will be invoked on the proposed
> transactions target. | +| `params` | `[]byte` | The parameters that will
> be passed to the method invocation on the proposed transactions target. |
> + +```go +func Propose(to Address, value TokenAmount, method string,
> params []byte) uint64 { + if !isSigner(msg.From) { + Fatal("not
> authorized") + }
> 
> I don't think thats useful. Then any random person can spam your multisig
> with bad transactions that we have to clear out, and can be misleading
> (making a typo on an approval transaction much more risky)
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub (
> https://github.com/filecoin-project/specs/pull/139#discussion_r269274406 )
> , or mute the thread (
> https://github.com/notifications/unsubscribe-auth/ADNBa5rr01S1oy4at-YjZQovNOoPCYzkks5vanJ0gaJpZM4a1bNp
> ).
> 
> 
>
